panics and fails in package tests when running with `count` flag.

### What are you trying to do?
Run tests with `go test -count=5 ./...` inside gonum on latest master branch

### What did you expect to happen?
All tests pass.

### What actually happened?

```
~/src/gonum$ time go test -count=5 ./...
?   	gonum.org/v1/gonum	[no test files]
?   	gonum.org/v1/gonum/blas	[no test files]
ok  	gonum.org/v1/gonum/blas/blas32	0.006s
ok  	gonum.org/v1/gonum/blas/blas64	0.025s
ok  	gonum.org/v1/gonum/blas/cblas128	13.900s
ok  	gonum.org/v1/gonum/blas/cblas64	0.007s
ok  	gonum.org/v1/gonum/blas/gonum	15.538s
ok  	gonum.org/v1/gonum/blas/testblas	0.045s
?   	gonum.org/v1/gonum/blas/testblas/benchautogen	[no test files]
ok  	gonum.org/v1/gonum/cmplxs	0.033s
ok  	gonum.org/v1/gonum/cmplxs/cscalar	0.030s
ok  	gonum.org/v1/gonum/diff/fd	0.041s
ok  	gonum.org/v1/gonum/dsp/fourier	266.819s
ok  	gonum.org/v1/gonum/dsp/fourier/internal/fftpack	50.642s
ok  	gonum.org/v1/gonum/dsp/window	0.024s
?   	gonum.org/v1/gonum/dsp/window/cmd/leakage	[no test files]
ok  	gonum.org/v1/gonum/floats	0.053s
ok  	gonum.org/v1/gonum/floats/scalar	0.046s
--- FAIL: TestCopy (0.00s)
panic: simple: node ID collision: 2 [recovered]
	panic: simple: node ID collision: 2

goroutine 35 [running]:
testing.tRunner.func1.2({0x51f7a0, 0xc000440b90})
	/usr/local/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1392 +0x39f
panic({0x51f7a0, 0xc000440b90})
	/usr/local/go/src/runtime/panic.go:838 +0x207
gonum.org/v1/gonum/graph/simple.(*UndirectedGraph).AddNode(0xc00000d920, {0x5781a0, 0x5775c8})
	/home/pato/src/gonum/graph/simple/undirected.go:47 +0x174
gonum.org/v1/gonum/graph.Copy({0x578b60, 0xc00000d920}, {0x578f28, 0xc00000d8f0})
	/home/pato/src/gonum/graph/graph.go:258 +0x6e
gonum.org/v1/gonum/graph_test.TestCopy(0xc00039a1a0)
	/home/pato/src/gonum/graph/graph_test.go:128 +0xdf
testing.tRunner(0xc00039a1a0, 0x552ec8)
	/usr/local/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1486 +0x35f
FAIL	gonum.org/v1/gonum/graph	0.046s
```

and I let it run some and it spewed out enough base64 data to overflow my terminal. 

```
... (continues for lots of lines)
sy779GVMQHBG6k5jDvI5A7/h8gdz+leJXgiJlPotgILcAIkNMmEKuilF2TCY6XlA4HH3cxpixsZpAAuOZggbyywQRaxwUxiAw2oV4DEXKjIllVMiagjvbz8Uxgml3VNi0sdbOjTGdwlqEcYY4Orgs1OjQ6IjOLC4O4/vR37xIkBGnr4GbQvQxGWOmBamrtcW6gkUXvVf3Fuhr09BAgomerO8IUl0k0C8QvDP6yIwxxhhjeslNzJg1xhhjjBkudm6MMcYYMyrs3BhjjDFmVNi5McYYY8yosHNjjDHGmFHx/wIAAP//26nG3JxnKyEAAAAASUVORK5CYII=
    eades_test.go:228: got node positions: map[int64]r2.Vec{0:r2.Vec{X:-1.626271083155058e+22, Y:5.903686807336693e+21}, 1:r2.Vec{X:1.8442322540405892e+23, Y:-6.698729760229611e+22}, 2:r2.Vec{X:-2.1820085261395386e+22, Y:7.926366132126938e+21}, 3:r2.Vec{X:1.3632640829871364e+23, Y:-4.9490471155818e+22}, 4:r2.Vec{X:-9.913379796171175e+23, Y:3.599273844604736e+23}, 5:r2.Vec{X:1.1550893845148759e+23, Y:-4.193332701353889e+22}, 6:r2.Vec{X:1.1550893845148766e+23, Y:-4.193332701354232e+22}, 7:r2.Vec{X:1.1550893845148693e+23, Y:-4.193332701354209e+22}, 8:r2.Vec{X:1.1550893845148647e+23, Y:-4.193332701354535e+22}, 9:r2.Vec{X:1.2331769410067117e+23, Y:-4.477318029382727e+22}, 10:r2.Vec{X:1.2331769410067117e+23, Y:-4.477318029382727e+22}}
FAIL
FAIL	gonum.org/v1/gonum/graph/layout	11.513s
ok  	gonum.org/v1/gonum/graph/multi	3.454s
ok  	gonum.org/v1/gonum/graph/network	0.092s
ok  	gonum.org/v1/gonum/graph/path	71.416s
ok  	gonum.org/v1/gonum/graph/path/dynamic	0.556s
ok  	gonum.org/v1/gonum/graph/path/internal/testgraphs	0.017s
ok  	gonum.org/v1/gonum/graph/product	10.435s
ok  	gonum.org/v1/gonum/graph/set/uid	38.160s
ok  	gonum.org/v1/gonum/graph/simple	1.809s
ok  	gonum.org/v1/gonum/graph/spectral	0.039s
ok  	gonum.org/v1/gonum/graph/testgraph	0.036s
ok  	gonum.org/v1/gonum/graph/topo	1.198s
ok  	gonum.org/v1/gonum/graph/traverse	0.722s
ok  	gonum.org/v1/gonum/integrate	0.496s
ok  	gonum.org/v1/gonum/integrate/quad	0.870s
?   	gonum.org/v1/gonum/integrate/testquad	[no test files]
ok  	gonum.org/v1/gonum/internal/asm/c128	0.054s
ok  	gonum.org/v1/gonum/internal/asm/c64	0.065s
ok  	gonum.org/v1/gonum/internal/asm/f32	0.078s
ok  	gonum.org/v1/gonum/internal/asm/f64	0.287s
ok  	gonum.org/v1/gonum/internal/cmplx64	0.003s
ok  	gonum.org/v1/gonum/internal/math32	0.029s
ok  	gonum.org/v1/gonum/internal/testrand	0.008s
ok  	gonum.org/v1/gonum/interp	0.020s
?   	gonum.org/v1/gonum/lapack	[no test files]
ok  	gonum.org/v1/gonum/lapack/gonum	466.612s
?   	gonum.org/v1/gonum/lapack/lapack64	[no test files]
ok  	gonum.org/v1/gonum/lapack/testlapack	0.051s
ok  	gonum.org/v1/gonum/mat	328.659s
ok  	gonum.org/v1/gonum/mathext	60.957s
ok  	gonum.org/v1/gonum/mathext/internal/amos	53.636s
?   	gonum.org/v1/gonum/mathext/internal/cephes	[no test files]
?   	gonum.org/v1/gonum/mathext/internal/gonum	[no test files]
ok  	gonum.org/v1/gonum/mathext/prng	0.071s
ok  	gonum.org/v1/gonum/num/dual	0.024s
ok  	gonum.org/v1/gonum/num/dualcmplx	0.051s
ok  	gonum.org/v1/gonum/num/dualquat	0.006s
ok  	gonum.org/v1/gonum/num/hyperdual	0.013s
ok  	gonum.org/v1/gonum/num/quat	0.020s
ok  	gonum.org/v1/gonum/optimize	147.554s
ok  	gonum.org/v1/gonum/optimize/convex/lp	351.519s
ok  	gonum.org/v1/gonum/optimize/functions	5.020s
ok  	gonum.org/v1/gonum/spatial/barneshut	530.937s
ok  	gonum.org/v1/gonum/spatial/kdtree	36.896s
?   	gonum.org/v1/gonum/spatial/r1	[no test files]
ok  	gonum.org/v1/gonum/spatial/r2	0.088s
ok  	gonum.org/v1/gonum/spatial/r3	0.057s
ok  	gonum.org/v1/gonum/spatial/vptree	51.064s
ok  	gonum.org/v1/gonum/stat	0.084s
ok  	gonum.org/v1/gonum/stat/card	107.157s
ok  	gonum.org/v1/gonum/stat/combin	0.426s
ok  	gonum.org/v1/gonum/stat/distmat	109.364s
ok  	gonum.org/v1/gonum/stat/distmv	154.774s
ok  	gonum.org/v1/gonum/stat/distuv	205.116s
ok  	gonum.org/v1/gonum/stat/mds	0.042s
--- FAIL: TestImportance (1.82s)
    sample_test.go:205: Covariance matrix mismatch
FAIL
FAIL	gonum.org/v1/gonum/stat/samplemv	46.678s
ok  	gonum.org/v1/gonum/stat/sampleuv	34.924s
ok  	gonum.org/v1/gonum/stat/spatial	3.608s
--- FAIL: TestOperations (0.00s)
    unit_test.go:213: value mismatch for test 0: got=5 want=3
    unit_test.go:213: value mismatch for test 1: got=4 want=2
    unit_test.go:210: dimension mismatch for test 2: got=4 m^2 want=2 m
    unit_test.go:213: value mismatch for test 2: got=4 want=2
    unit_test.go:213: value mismatch for test 3: got=4 want=2
    unit_test.go:210: dimension mismatch for test 4: got=0.25 m^-2 want=0.5 m^-1
    unit_test.go:213: value mismatch for test 4: got=0.25 want=0.5
    unit_test.go:213: value mismatch for test 5: got=0.25 want=0.5
--- FAIL: TestOperations (0.00s)
    unit_test.go:213: value mismatch for test 0: got=7 want=3
    unit_test.go:213: value mismatch for test 1: got=8 want=2
    unit_test.go:210: dimension mismatch for test 2: got=8 m^3 want=2 m
    unit_test.go:213: value mismatch for test 2: got=8 want=2
    unit_test.go:213: value mismatch for test 3: got=8 want=2
    unit_test.go:210: dimension mismatch for test 4: got=0.125 m^-3 want=0.5 m^-1
    unit_test.go:213: value mismatch for test 4: got=0.125 want=0.5
    unit_test.go:213: value mismatch for test 5: got=0.125 want=0.5
--- FAIL: TestOperations (0.00s)
    unit_test.go:213: value mismatch for test 0: got=9 want=3
    unit_test.go:213: value mismatch for test 1: got=16 want=2
    unit_test.go:210: dimension mismatch for test 2: got=16 m^4 want=2 m
    unit_test.go:213: value mismatch for test 2: got=16 want=2
    unit_test.go:213: value mismatch for test 3: got=16 want=2
    unit_test.go:210: dimension mismatch for test 4: got=0.0625 m^-4 want=0.5 m^-1
    unit_test.go:213: value mismatch for test 4: got=0.0625 want=0.5
    unit_test.go:213: value mismatch for test 5: got=0.0625 want=0.5
--- FAIL: TestOperations (0.00s)
    unit_test.go:213: value mismatch for test 0: got=11 want=3
    unit_test.go:213: value mismatch for test 1: got=32 want=2
    unit_test.go:210: dimension mismatch for test 2: got=32 m^5 want=2 m
    unit_test.go:213: value mismatch for test 2: got=32 want=2
    unit_test.go:213: value mismatch for test 3: got=32 want=2
    unit_test.go:210: dimension mismatch for test 4: got=0.03125 m^-5 want=0.5 m^-1
    unit_test.go:213: value mismatch for test 4: got=0.03125 want=0.5
    unit_test.go:213: value mismatch for test 5: got=0.03125 want=0.5
FAIL
FAIL	gonum.org/v1/gonum/unit	0.069s
ok  	gonum.org/v1/gonum/unit/constant	0.037s
FAIL

real	14m28,957s
user	101m43,714s
sys	2m29,247s
```

### What version of Go and Gonum are you using?
```
$ go version
go version go1.18.1 linux/amd64
```

### Does this issue reproduce with the current master?
Yes.

--- a/graph/graph_test.go
+++ b/graph/graph_test.go
@@ -17,113 +17,113 @@ type graphBuilder interface {
 	graph.Builder
 }
 
-var copyTests = []struct {
-	desc string
-
-	src graph.Graph
-	dst graphBuilder
-
-	// If want is nil, compare to src.
-	want graph.Graph
-}{
-	{
-		desc: "undirected to undirected",
-		src: func() graph.Graph {
-			g := simple.NewUndirectedGraph()
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-			} {
-				g.SetEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewUndirectedGraph(),
-	},
-	{
-		desc: "undirected to directed",
-		src: func() graph.Graph {
-			g := simple.NewUndirectedGraph()
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-			} {
-				g.SetEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewDirectedGraph(),
-
-		want: func() graph.Graph {
-			g := simple.NewDirectedGraph()
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-			} {
-				// want is a directed graph copied from
-				// an undirected graph so we need to have
-				// all edges in both directions.
-				g.SetEdge(e)
-				e.T, e.F = e.F, e.T
-				g.SetEdge(e)
-			}
-			return g
-		}(),
-	},
-	{
-		desc: "directed to undirected",
-		src: func() graph.Graph {
-			g := simple.NewDirectedGraph()
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-			} {
-				g.SetEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewUndirectedGraph(),
-
-		want: func() graph.Graph {
-			g := simple.NewUndirectedGraph()
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-			} {
-				g.SetEdge(e)
-			}
-			return g
-		}(),
-	},
-	{
-		desc: "directed to directed",
-		src: func() graph.Graph {
-			g := simple.NewDirectedGraph()
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-			} {
-				g.SetEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewDirectedGraph(),
-	},
-}
-
 func TestCopy(t *testing.T) {
+	copyTests := []struct {
+		desc string
+
+		src graph.Graph
+		dst graphBuilder
+
+		// If want is nil, compare to src.
+		want graph.Graph
+	}{
+		{
+			desc: "undirected to undirected",
+			src: func() graph.Graph {
+				g := simple.NewUndirectedGraph()
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+				} {
+					g.SetEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewUndirectedGraph(),
+		},
+		{
+			desc: "undirected to directed",
+			src: func() graph.Graph {
+				g := simple.NewUndirectedGraph()
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+				} {
+					g.SetEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewDirectedGraph(),
+
+			want: func() graph.Graph {
+				g := simple.NewDirectedGraph()
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+				} {
+					// want is a directed graph copied from
+					// an undirected graph so we need to have
+					// all edges in both directions.
+					g.SetEdge(e)
+					e.T, e.F = e.F, e.T
+					g.SetEdge(e)
+				}
+				return g
+			}(),
+		},
+		{
+			desc: "directed to undirected",
+			src: func() graph.Graph {
+				g := simple.NewDirectedGraph()
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+				} {
+					g.SetEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewUndirectedGraph(),
+
+			want: func() graph.Graph {
+				g := simple.NewUndirectedGraph()
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+				} {
+					g.SetEdge(e)
+				}
+				return g
+			}(),
+		},
+		{
+			desc: "directed to directed",
+			src: func() graph.Graph {
+				g := simple.NewDirectedGraph()
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+				} {
+					g.SetEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewDirectedGraph(),
+		},
+	}
+
 	for _, test := range copyTests {
 		graph.Copy(test.dst, test.src)
 		want := test.want
@@ -141,113 +141,113 @@ type graphWeightedBuilder interface {
 	graph.WeightedBuilder
 }
 
-var copyWeightedTests = []struct {
-	desc string
-
-	src graph.Weighted
-	dst graphWeightedBuilder
-
-	// If want is nil, compare to src.
-	want graph.Graph
-}{
-	{
-		desc: "undirected to undirected",
-		src: func() graph.Weighted {
-			g := simple.NewWeightedUndirectedGraph(0, 0)
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.WeightedEdge{
-				{F: simple.Node(0), T: simple.Node(1), W: 1},
-				{F: simple.Node(0), T: simple.Node(3), W: 1},
-				{F: simple.Node(1), T: simple.Node(2), W: 1},
-			} {
-				g.SetWeightedEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewWeightedUndirectedGraph(0, 0),
-	},
-	{
-		desc: "undirected to directed",
-		src: func() graph.Weighted {
-			g := simple.NewWeightedUndirectedGraph(0, 0)
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.WeightedEdge{
-				{F: simple.Node(0), T: simple.Node(1), W: 1},
-				{F: simple.Node(0), T: simple.Node(3), W: 1},
-				{F: simple.Node(1), T: simple.Node(2), W: 1},
-			} {
-				g.SetWeightedEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewWeightedDirectedGraph(0, 0),
-
-		want: func() graph.Graph {
-			g := simple.NewWeightedDirectedGraph(0, 0)
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.WeightedEdge{
-				{F: simple.Node(0), T: simple.Node(1), W: 1},
-				{F: simple.Node(0), T: simple.Node(3), W: 1},
-				{F: simple.Node(1), T: simple.Node(2), W: 1},
-			} {
-				// want is a directed graph copied from
-				// an undirected graph so we need to have
-				// all edges in both directions.
-				g.SetWeightedEdge(e)
-				e.T, e.F = e.F, e.T
-				g.SetWeightedEdge(e)
-			}
-			return g
-		}(),
-	},
-	{
-		desc: "directed to undirected",
-		src: func() graph.Weighted {
-			g := simple.NewWeightedDirectedGraph(0, 0)
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.WeightedEdge{
-				{F: simple.Node(0), T: simple.Node(1), W: 1},
-				{F: simple.Node(0), T: simple.Node(3), W: 1},
-				{F: simple.Node(1), T: simple.Node(2), W: 1},
-			} {
-				g.SetWeightedEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewWeightedUndirectedGraph(0, 0),
-
-		want: func() graph.Weighted {
-			g := simple.NewWeightedUndirectedGraph(0, 0)
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.WeightedEdge{
-				{F: simple.Node(0), T: simple.Node(1), W: 1},
-				{F: simple.Node(0), T: simple.Node(3), W: 1},
-				{F: simple.Node(1), T: simple.Node(2), W: 1},
-			} {
-				g.SetWeightedEdge(e)
-			}
-			return g
-		}(),
-	},
-	{
-		desc: "directed to directed",
-		src: func() graph.Weighted {
-			g := simple.NewWeightedDirectedGraph(0, 0)
-			g.AddNode(simple.Node(-1))
-			for _, e := range []simple.WeightedEdge{
-				{F: simple.Node(0), T: simple.Node(1), W: 1},
-				{F: simple.Node(0), T: simple.Node(3), W: 1},
-				{F: simple.Node(1), T: simple.Node(2), W: 1},
-			} {
-				g.SetWeightedEdge(e)
-			}
-			return g
-		}(),
-		dst: simple.NewWeightedDirectedGraph(0, 0),
-	},
-}
-
 func TestCopyWeighted(t *testing.T) {
+	copyWeightedTests := []struct {
+		desc string
+
+		src graph.Weighted
+		dst graphWeightedBuilder
+
+		// If want is nil, compare to src.
+		want graph.Graph
+	}{
+		{
+			desc: "undirected to undirected",
+			src: func() graph.Weighted {
+				g := simple.NewWeightedUndirectedGraph(0, 0)
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.WeightedEdge{
+					{F: simple.Node(0), T: simple.Node(1), W: 1},
+					{F: simple.Node(0), T: simple.Node(3), W: 1},
+					{F: simple.Node(1), T: simple.Node(2), W: 1},
+				} {
+					g.SetWeightedEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewWeightedUndirectedGraph(0, 0),
+		},
+		{
+			desc: "undirected to directed",
+			src: func() graph.Weighted {
+				g := simple.NewWeightedUndirectedGraph(0, 0)
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.WeightedEdge{
+					{F: simple.Node(0), T: simple.Node(1), W: 1},
+					{F: simple.Node(0), T: simple.Node(3), W: 1},
+					{F: simple.Node(1), T: simple.Node(2), W: 1},
+				} {
+					g.SetWeightedEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewWeightedDirectedGraph(0, 0),
+
+			want: func() graph.Graph {
+				g := simple.NewWeightedDirectedGraph(0, 0)
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.WeightedEdge{
+					{F: simple.Node(0), T: simple.Node(1), W: 1},
+					{F: simple.Node(0), T: simple.Node(3), W: 1},
+					{F: simple.Node(1), T: simple.Node(2), W: 1},
+				} {
+					// want is a directed graph copied from
+					// an undirected graph so we need to have
+					// all edges in both directions.
+					g.SetWeightedEdge(e)
+					e.T, e.F = e.F, e.T
+					g.SetWeightedEdge(e)
+				}
+				return g
+			}(),
+		},
+		{
+			desc: "directed to undirected",
+			src: func() graph.Weighted {
+				g := simple.NewWeightedDirectedGraph(0, 0)
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.WeightedEdge{
+					{F: simple.Node(0), T: simple.Node(1), W: 1},
+					{F: simple.Node(0), T: simple.Node(3), W: 1},
+					{F: simple.Node(1), T: simple.Node(2), W: 1},
+				} {
+					g.SetWeightedEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewWeightedUndirectedGraph(0, 0),
+
+			want: func() graph.Weighted {
+				g := simple.NewWeightedUndirectedGraph(0, 0)
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.WeightedEdge{
+					{F: simple.Node(0), T: simple.Node(1), W: 1},
+					{F: simple.Node(0), T: simple.Node(3), W: 1},
+					{F: simple.Node(1), T: simple.Node(2), W: 1},
+				} {
+					g.SetWeightedEdge(e)
+				}
+				return g
+			}(),
+		},
+		{
+			desc: "directed to directed",
+			src: func() graph.Weighted {
+				g := simple.NewWeightedDirectedGraph(0, 0)
+				g.AddNode(simple.Node(-1))
+				for _, e := range []simple.WeightedEdge{
+					{F: simple.Node(0), T: simple.Node(1), W: 1},
+					{F: simple.Node(0), T: simple.Node(3), W: 1},
+					{F: simple.Node(1), T: simple.Node(2), W: 1},
+				} {
+					g.SetWeightedEdge(e)
+				}
+				return g
+			}(),
+			dst: simple.NewWeightedDirectedGraph(0, 0),
+		},
+	}
+
 	for _, test := range copyWeightedTests {
 		graph.CopyWeighted(test.dst, test.src)
 		want := test.want

--- a/graph/layout/eades_test.go
+++ b/graph/layout/eades_test.go
@@ -19,184 +19,184 @@ import (
 	. "gonum.org/v1/gonum/graph/layout"
 )
 
-var eadesR2Tests = []struct {
-	name      string
-	g         graph.Graph
-	param     EadesR2
-	wantIters int
-}{
-	{
-		name: "line",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-		param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
-		wantIters: 100,
-	},
-	{
-		name: "square",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(3)},
-				{F: simple.Node(2), T: simple.Node(3)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-		param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
-		wantIters: 100,
-	},
-	{
-		name: "tetrahedron",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(2)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(3)},
-				{F: simple.Node(2), T: simple.Node(3)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-		param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
-		wantIters: 100,
-	},
-	{
-		name: "sheet",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(4)},
-				{F: simple.Node(2), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(4)},
-				{F: simple.Node(3), T: simple.Node(6)},
-				{F: simple.Node(4), T: simple.Node(5)},
-				{F: simple.Node(4), T: simple.Node(7)},
-				{F: simple.Node(5), T: simple.Node(8)},
-				{F: simple.Node(6), T: simple.Node(7)},
-				{F: simple.Node(7), T: simple.Node(8)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-		param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
-		wantIters: 100,
-	},
-	{
-		name: "tube",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(2)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(4)},
-				{F: simple.Node(2), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(4)},
-				{F: simple.Node(3), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(6)},
-				{F: simple.Node(4), T: simple.Node(5)},
-				{F: simple.Node(4), T: simple.Node(7)},
-				{F: simple.Node(5), T: simple.Node(8)},
-				{F: simple.Node(6), T: simple.Node(7)},
-				{F: simple.Node(6), T: simple.Node(8)},
-				{F: simple.Node(7), T: simple.Node(8)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-		param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
-		wantIters: 100,
-	},
-	{
-		// This test does not produce a good layout, but is here to
-		// ensure that Update does not panic with steep decent rates.
-		name: "tube-steep",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(2)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(4)},
-				{F: simple.Node(2), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(4)},
-				{F: simple.Node(3), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(6)},
-				{F: simple.Node(4), T: simple.Node(5)},
-				{F: simple.Node(4), T: simple.Node(7)},
-				{F: simple.Node(5), T: simple.Node(8)},
-				{F: simple.Node(6), T: simple.Node(7)},
-				{F: simple.Node(6), T: simple.Node(8)},
-				{F: simple.Node(7), T: simple.Node(8)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-		param:     EadesR2{Repulsion: 1, Rate: 1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
-		wantIters: 99,
-	},
-
-	{
-		name: "wp_page", // https://en.wikipedia.org/wiki/PageRank#/media/File:PageRanks-Example.jpg
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(4)},
-				{F: simple.Node(1), T: simple.Node(5)},
-				{F: simple.Node(1), T: simple.Node(6)},
-				{F: simple.Node(1), T: simple.Node(7)},
-				{F: simple.Node(1), T: simple.Node(8)},
-				{F: simple.Node(3), T: simple.Node(4)},
-				{F: simple.Node(4), T: simple.Node(5)},
-				{F: simple.Node(4), T: simple.Node(6)},
-				{F: simple.Node(4), T: simple.Node(7)},
-				{F: simple.Node(4), T: simple.Node(8)},
-				{F: simple.Node(4), T: simple.Node(9)},
-				{F: simple.Node(4), T: simple.Node(10)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-		param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
-		wantIters: 100,
-	},
-}
-
 func TestEadesR2(t *testing.T) {
+	eadesR2Tests := []struct {
+		name      string
+		g         graph.Graph
+		param     EadesR2
+		wantIters int
+	}{
+		{
+			name: "line",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+			param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
+			wantIters: 100,
+		},
+		{
+			name: "square",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(3)},
+					{F: simple.Node(2), T: simple.Node(3)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+			param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
+			wantIters: 100,
+		},
+		{
+			name: "tetrahedron",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(2)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(3)},
+					{F: simple.Node(2), T: simple.Node(3)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+			param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
+			wantIters: 100,
+		},
+		{
+			name: "sheet",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(4)},
+					{F: simple.Node(2), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(4)},
+					{F: simple.Node(3), T: simple.Node(6)},
+					{F: simple.Node(4), T: simple.Node(5)},
+					{F: simple.Node(4), T: simple.Node(7)},
+					{F: simple.Node(5), T: simple.Node(8)},
+					{F: simple.Node(6), T: simple.Node(7)},
+					{F: simple.Node(7), T: simple.Node(8)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+			param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
+			wantIters: 100,
+		},
+		{
+			name: "tube",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(2)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(4)},
+					{F: simple.Node(2), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(4)},
+					{F: simple.Node(3), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(6)},
+					{F: simple.Node(4), T: simple.Node(5)},
+					{F: simple.Node(4), T: simple.Node(7)},
+					{F: simple.Node(5), T: simple.Node(8)},
+					{F: simple.Node(6), T: simple.Node(7)},
+					{F: simple.Node(6), T: simple.Node(8)},
+					{F: simple.Node(7), T: simple.Node(8)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+			param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
+			wantIters: 100,
+		},
+		{
+			// This test does not produce a good layout, but is here to
+			// ensure that Update does not panic with steep decent rates.
+			name: "tube-steep",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(2)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(4)},
+					{F: simple.Node(2), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(4)},
+					{F: simple.Node(3), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(6)},
+					{F: simple.Node(4), T: simple.Node(5)},
+					{F: simple.Node(4), T: simple.Node(7)},
+					{F: simple.Node(5), T: simple.Node(8)},
+					{F: simple.Node(6), T: simple.Node(7)},
+					{F: simple.Node(6), T: simple.Node(8)},
+					{F: simple.Node(7), T: simple.Node(8)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+			param:     EadesR2{Repulsion: 1, Rate: 1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
+			wantIters: 99,
+		},
+
+		{
+			name: "wp_page", // https://en.wikipedia.org/wiki/PageRank#/media/File:PageRanks-Example.jpg
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(4)},
+					{F: simple.Node(1), T: simple.Node(5)},
+					{F: simple.Node(1), T: simple.Node(6)},
+					{F: simple.Node(1), T: simple.Node(7)},
+					{F: simple.Node(1), T: simple.Node(8)},
+					{F: simple.Node(3), T: simple.Node(4)},
+					{F: simple.Node(4), T: simple.Node(5)},
+					{F: simple.Node(4), T: simple.Node(6)},
+					{F: simple.Node(4), T: simple.Node(7)},
+					{F: simple.Node(4), T: simple.Node(8)},
+					{F: simple.Node(4), T: simple.Node(9)},
+					{F: simple.Node(4), T: simple.Node(10)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+			param:     EadesR2{Repulsion: 1, Rate: 0.1, Updates: 100, Theta: 0.1, Src: rand.NewSource(1)},
+			wantIters: 100,
+		},
+	}
+
 	for _, test := range eadesR2Tests {
 		eades := test.param
 		o := NewOptimizerR2(test.g, eades.Update)

--- a/graph/layout/isomap_test.go
+++ b/graph/layout/isomap_test.go
@@ -28,138 +28,138 @@ var (
 	arch string
 )
 
-var isomapR2Tests = []struct {
-	name string
-	g    graph.Graph
-}{
-	{
-		name: "line_isomap",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-	},
-	{
-		name: "square_isomap",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(3)},
-				{F: simple.Node(2), T: simple.Node(3)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-	},
-	{
-		name: "tetrahedron_isomap",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(2)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(3)},
-				{F: simple.Node(2), T: simple.Node(3)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-	},
-	{
-		name: "sheet_isomap",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(4)},
-				{F: simple.Node(2), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(4)},
-				{F: simple.Node(3), T: simple.Node(6)},
-				{F: simple.Node(4), T: simple.Node(5)},
-				{F: simple.Node(4), T: simple.Node(7)},
-				{F: simple.Node(5), T: simple.Node(8)},
-				{F: simple.Node(6), T: simple.Node(7)},
-				{F: simple.Node(7), T: simple.Node(8)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-	},
-	{
-		name: "tube_isomap",
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(1)},
-				{F: simple.Node(0), T: simple.Node(2)},
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(4)},
-				{F: simple.Node(2), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(4)},
-				{F: simple.Node(3), T: simple.Node(5)},
-				{F: simple.Node(3), T: simple.Node(6)},
-				{F: simple.Node(4), T: simple.Node(5)},
-				{F: simple.Node(4), T: simple.Node(7)},
-				{F: simple.Node(5), T: simple.Node(8)},
-				{F: simple.Node(6), T: simple.Node(7)},
-				{F: simple.Node(6), T: simple.Node(8)},
-				{F: simple.Node(7), T: simple.Node(8)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-	},
-	{
-		name: "wp_page_isomap", // https://en.wikipedia.org/wiki/PageRank#/media/File:PageRanks-Example.jpg
-		g: func() graph.Graph {
-			edges := []simple.Edge{
-				{F: simple.Node(0), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(2)},
-				{F: simple.Node(1), T: simple.Node(3)},
-				{F: simple.Node(1), T: simple.Node(4)},
-				{F: simple.Node(1), T: simple.Node(5)},
-				{F: simple.Node(1), T: simple.Node(6)},
-				{F: simple.Node(1), T: simple.Node(7)},
-				{F: simple.Node(1), T: simple.Node(8)},
-				{F: simple.Node(3), T: simple.Node(4)},
-				{F: simple.Node(4), T: simple.Node(5)},
-				{F: simple.Node(4), T: simple.Node(6)},
-				{F: simple.Node(4), T: simple.Node(7)},
-				{F: simple.Node(4), T: simple.Node(8)},
-				{F: simple.Node(4), T: simple.Node(9)},
-				{F: simple.Node(4), T: simple.Node(10)},
-			}
-			g := simple.NewUndirectedGraph()
-			for _, e := range edges {
-				g.SetEdge(e)
-			}
-			return orderedGraph{g}
-		}(),
-	},
-}
-
 func TestIsomapR2(t *testing.T) {
+	isomapR2Tests := []struct {
+		name string
+		g    graph.Graph
+	}{
+		{
+			name: "line_isomap",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+		},
+		{
+			name: "square_isomap",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(3)},
+					{F: simple.Node(2), T: simple.Node(3)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+		},
+		{
+			name: "tetrahedron_isomap",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(2)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(3)},
+					{F: simple.Node(2), T: simple.Node(3)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+		},
+		{
+			name: "sheet_isomap",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(4)},
+					{F: simple.Node(2), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(4)},
+					{F: simple.Node(3), T: simple.Node(6)},
+					{F: simple.Node(4), T: simple.Node(5)},
+					{F: simple.Node(4), T: simple.Node(7)},
+					{F: simple.Node(5), T: simple.Node(8)},
+					{F: simple.Node(6), T: simple.Node(7)},
+					{F: simple.Node(7), T: simple.Node(8)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+		},
+		{
+			name: "tube_isomap",
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(1)},
+					{F: simple.Node(0), T: simple.Node(2)},
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(4)},
+					{F: simple.Node(2), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(4)},
+					{F: simple.Node(3), T: simple.Node(5)},
+					{F: simple.Node(3), T: simple.Node(6)},
+					{F: simple.Node(4), T: simple.Node(5)},
+					{F: simple.Node(4), T: simple.Node(7)},
+					{F: simple.Node(5), T: simple.Node(8)},
+					{F: simple.Node(6), T: simple.Node(7)},
+					{F: simple.Node(6), T: simple.Node(8)},
+					{F: simple.Node(7), T: simple.Node(8)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+		},
+		{
+			name: "wp_page_isomap", // https://en.wikipedia.org/wiki/PageRank#/media/File:PageRanks-Example.jpg
+			g: func() graph.Graph {
+				edges := []simple.Edge{
+					{F: simple.Node(0), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(2)},
+					{F: simple.Node(1), T: simple.Node(3)},
+					{F: simple.Node(1), T: simple.Node(4)},
+					{F: simple.Node(1), T: simple.Node(5)},
+					{F: simple.Node(1), T: simple.Node(6)},
+					{F: simple.Node(1), T: simple.Node(7)},
+					{F: simple.Node(1), T: simple.Node(8)},
+					{F: simple.Node(3), T: simple.Node(4)},
+					{F: simple.Node(4), T: simple.Node(5)},
+					{F: simple.Node(4), T: simple.Node(6)},
+					{F: simple.Node(4), T: simple.Node(7)},
+					{F: simple.Node(4), T: simple.Node(8)},
+					{F: simple.Node(4), T: simple.Node(9)},
+					{F: simple.Node(4), T: simple.Node(10)},
+				}
+				g := simple.NewUndirectedGraph()
+				for _, e := range edges {
+					g.SetEdge(e)
+				}
+				return orderedGraph{g}
+			}(),
+		},
+	}
+
 	for _, test := range isomapR2Tests {
 		o := NewOptimizerR2(test.g, IsomapR2{}.Update)
 		var n int

--- a/graph/nodes_edges_test.go
+++ b/graph/nodes_edges_test.go
@@ -21,48 +21,48 @@ import (
 // weightedlines
 // empty
 
-var nodesOfTests = []struct {
-	name  string
-	nodes graph.Nodes
-	want  []graph.Node
-}{
-	{
-		name:  "nil",
-		nodes: nil,
-		want:  nil,
-	},
-	{
-		name:  "empty",
-		nodes: graph.Empty,
-		want:  nil,
-	},
-	{
-		name:  "no nodes",
-		nodes: iterator.NewOrderedNodes(nil),
-		want:  nil,
-	},
-	{
-		name:  "implicit nodes",
-		nodes: iterator.NewImplicitNodes(-1, 4, func(id int) graph.Node { return simple.Node(id) }),
-		want:  []graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)},
-	},
-	{
-		name:  "no slice method",
-		nodes: basicNodes{iterator.NewOrderedNodes([]graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)})},
-		want:  []graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)},
-	},
-	{
-		name:  "explicit nodes",
-		nodes: iterator.NewOrderedNodes([]graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)}),
-		want:  []graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)},
-	},
-}
-
 type basicNodes struct {
 	graph.Nodes
 }
 
 func TestNodesOf(t *testing.T) {
+	nodesOfTests := []struct {
+		name  string
+		nodes graph.Nodes
+		want  []graph.Node
+	}{
+		{
+			name:  "nil",
+			nodes: nil,
+			want:  nil,
+		},
+		{
+			name:  "empty",
+			nodes: graph.Empty,
+			want:  nil,
+		},
+		{
+			name:  "no nodes",
+			nodes: iterator.NewOrderedNodes(nil),
+			want:  nil,
+		},
+		{
+			name:  "implicit nodes",
+			nodes: iterator.NewImplicitNodes(-1, 4, func(id int) graph.Node { return simple.Node(id) }),
+			want:  []graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)},
+		},
+		{
+			name:  "no slice method",
+			nodes: basicNodes{iterator.NewOrderedNodes([]graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)})},
+			want:  []graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)},
+		},
+		{
+			name:  "explicit nodes",
+			nodes: iterator.NewOrderedNodes([]graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)}),
+			want:  []graph.Node{simple.Node(-1), simple.Node(0), simple.Node(1), simple.Node(2), simple.Node(3)},
+		},
+	}
+
 	for _, test := range nodesOfTests {
 		got := graph.NodesOf(test.nodes)
 		if !reflect.DeepEqual(got, test.want) {
@@ -71,59 +71,59 @@ func TestNodesOf(t *testing.T) {
 	}
 }
 
-var edgesOfTests = []struct {
-	name  string
-	edges graph.Edges
-	want  []graph.Edge
-}{
-	{
-		name:  "nil",
-		edges: nil,
-		want:  nil,
-	},
-	{
-		name:  "empty",
-		edges: graph.Empty,
-		want:  nil,
-	},
-	{
-		name:  "no edges",
-		edges: iterator.NewOrderedEdges(nil),
-		want:  nil,
-	},
-	{
-		name: "no slice method",
-		edges: basicEdges{iterator.NewOrderedEdges([]graph.Edge{
-			simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
-			simple.Edge{F: simple.Node(1), T: simple.Node(2)},
-			simple.Edge{F: simple.Node(3), T: simple.Node(4)},
-		})},
-		want: []graph.Edge{
-			simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
-			simple.Edge{F: simple.Node(1), T: simple.Node(2)},
-			simple.Edge{F: simple.Node(3), T: simple.Node(4)},
-		},
-	},
-	{
-		name: "explicit edges",
-		edges: iterator.NewOrderedEdges([]graph.Edge{
-			simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
-			simple.Edge{F: simple.Node(1), T: simple.Node(2)},
-			simple.Edge{F: simple.Node(3), T: simple.Node(4)},
-		}),
-		want: []graph.Edge{
-			simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
-			simple.Edge{F: simple.Node(1), T: simple.Node(2)},
-			simple.Edge{F: simple.Node(3), T: simple.Node(4)},
-		},
-	},
-}
-
 type basicEdges struct {
 	graph.Edges
 }
 
 func TestEdgesOf(t *testing.T) {
+	edgesOfTests := []struct {
+		name  string
+		edges graph.Edges
+		want  []graph.Edge
+	}{
+		{
+			name:  "nil",
+			edges: nil,
+			want:  nil,
+		},
+		{
+			name:  "empty",
+			edges: graph.Empty,
+			want:  nil,
+		},
+		{
+			name:  "no edges",
+			edges: iterator.NewOrderedEdges(nil),
+			want:  nil,
+		},
+		{
+			name: "no slice method",
+			edges: basicEdges{iterator.NewOrderedEdges([]graph.Edge{
+				simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
+				simple.Edge{F: simple.Node(1), T: simple.Node(2)},
+				simple.Edge{F: simple.Node(3), T: simple.Node(4)},
+			})},
+			want: []graph.Edge{
+				simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
+				simple.Edge{F: simple.Node(1), T: simple.Node(2)},
+				simple.Edge{F: simple.Node(3), T: simple.Node(4)},
+			},
+		},
+		{
+			name: "explicit edges",
+			edges: iterator.NewOrderedEdges([]graph.Edge{
+				simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
+				simple.Edge{F: simple.Node(1), T: simple.Node(2)},
+				simple.Edge{F: simple.Node(3), T: simple.Node(4)},
+			}),
+			want: []graph.Edge{
+				simple.Edge{F: simple.Node(-1), T: simple.Node(0)},
+				simple.Edge{F: simple.Node(1), T: simple.Node(2)},
+				simple.Edge{F: simple.Node(3), T: simple.Node(4)},
+			},
+		},
+	}
+
 	for _, test := range edgesOfTests {
 		got := graph.EdgesOf(test.edges)
 		if !reflect.DeepEqual(got, test.want) {
@@ -132,59 +132,59 @@ func TestEdgesOf(t *testing.T) {
 	}
 }
 
-var weightedEdgesOfTests = []struct {
-	name  string
-	edges graph.WeightedEdges
-	want  []graph.WeightedEdge
-}{
-	{
-		name:  "nil",
-		edges: nil,
-		want:  nil,
-	},
-	{
-		name:  "empty",
-		edges: graph.Empty,
-		want:  nil,
-	},
-	{
-		name:  "no edges",
-		edges: iterator.NewOrderedWeightedEdges(nil),
-		want:  nil,
-	},
-	{
-		name: "no slice method",
-		edges: basicWeightedEdges{iterator.NewOrderedWeightedEdges([]graph.WeightedEdge{
-			simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
-			simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
-			simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
-		})},
-		want: []graph.WeightedEdge{
-			simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
-			simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
-			simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
-		},
-	},
-	{
-		name: "explicit edges",
-		edges: iterator.NewOrderedWeightedEdges([]graph.WeightedEdge{
-			simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
-			simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
-			simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
-		}),
-		want: []graph.WeightedEdge{
-			simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
-			simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
-			simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
-		},
-	},
-}
-
 type basicWeightedEdges struct {
 	graph.WeightedEdges
 }
 
 func TestWeightedEdgesOf(t *testing.T) {
+	weightedEdgesOfTests := []struct {
+		name  string
+		edges graph.WeightedEdges
+		want  []graph.WeightedEdge
+	}{
+		{
+			name:  "nil",
+			edges: nil,
+			want:  nil,
+		},
+		{
+			name:  "empty",
+			edges: graph.Empty,
+			want:  nil,
+		},
+		{
+			name:  "no edges",
+			edges: iterator.NewOrderedWeightedEdges(nil),
+			want:  nil,
+		},
+		{
+			name: "no slice method",
+			edges: basicWeightedEdges{iterator.NewOrderedWeightedEdges([]graph.WeightedEdge{
+				simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
+				simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
+				simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
+			})},
+			want: []graph.WeightedEdge{
+				simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
+				simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
+				simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
+			},
+		},
+		{
+			name: "explicit edges",
+			edges: iterator.NewOrderedWeightedEdges([]graph.WeightedEdge{
+				simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
+				simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
+				simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
+			}),
+			want: []graph.WeightedEdge{
+				simple.WeightedEdge{F: simple.Node(-1), T: simple.Node(0), W: 1},
+				simple.WeightedEdge{F: simple.Node(1), T: simple.Node(2), W: 2},
+				simple.WeightedEdge{F: simple.Node(3), T: simple.Node(4), W: 3},
+			},
+		},
+	}
+
 	for _, test := range weightedEdgesOfTests {
 		got := graph.WeightedEdgesOf(test.edges)
 		if !reflect.DeepEqual(got, test.want) {
@@ -193,59 +193,59 @@ func TestWeightedEdgesOf(t *testing.T) {
 	}
 }
 
-var linesOfTests = []struct {
-	name  string
-	lines graph.Lines
-	want  []graph.Line
-}{
-	{
-		name:  "nil",
-		lines: nil,
-		want:  nil,
-	},
-	{
-		name:  "empty",
-		lines: graph.Empty,
-		want:  nil,
-	},
-	{
-		name:  "no edges",
-		lines: iterator.NewOrderedLines(nil),
-		want:  nil,
-	},
-	{
-		name: "no slice method",
-		lines: basicLines{iterator.NewOrderedLines([]graph.Line{
-			multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
-			multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
-			multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
-		})},
-		want: []graph.Line{
-			multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
-			multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
-			multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
-		},
-	},
-	{
-		name: "explicit edges",
-		lines: iterator.NewOrderedLines([]graph.Line{
-			multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
-			multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
-			multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
-		}),
-		want: []graph.Line{
-			multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
-			multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
-			multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
-		},
-	},
-}
-
 type basicLines struct {
 	graph.Lines
 }
 
 func TestLinesOf(t *testing.T) {
+	linesOfTests := []struct {
+		name  string
+		lines graph.Lines
+		want  []graph.Line
+	}{
+		{
+			name:  "nil",
+			lines: nil,
+			want:  nil,
+		},
+		{
+			name:  "empty",
+			lines: graph.Empty,
+			want:  nil,
+		},
+		{
+			name:  "no edges",
+			lines: iterator.NewOrderedLines(nil),
+			want:  nil,
+		},
+		{
+			name: "no slice method",
+			lines: basicLines{iterator.NewOrderedLines([]graph.Line{
+				multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
+				multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
+				multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
+			})},
+			want: []graph.Line{
+				multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
+				multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
+				multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
+			},
+		},
+		{
+			name: "explicit edges",
+			lines: iterator.NewOrderedLines([]graph.Line{
+				multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
+				multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
+				multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
+			}),
+			want: []graph.Line{
+				multi.Line{F: multi.Node(-1), T: multi.Node(0), UID: -1},
+				multi.Line{F: multi.Node(1), T: multi.Node(2), UID: 0},
+				multi.Line{F: multi.Node(3), T: multi.Node(4), UID: 1},
+			},
+		},
+	}
+
 	for _, test := range linesOfTests {
 		got := graph.LinesOf(test.lines)
 		if !reflect.DeepEqual(got, test.want) {
@@ -254,59 +254,59 @@ func TestLinesOf(t *testing.T) {
 	}
 }
 
-var weightedLinesOfTests = []struct {
-	name  string
-	lines graph.WeightedLines
-	want  []graph.WeightedLine
-}{
-	{
-		name:  "nil",
-		lines: nil,
-		want:  nil,
-	},
-	{
-		name:  "empty",
-		lines: graph.Empty,
-		want:  nil,
-	},
-	{
-		name:  "no edges",
-		lines: iterator.NewOrderedWeightedLines(nil),
-		want:  nil,
-	},
-	{
-		name: "no slice method",
-		lines: basicWeightedLines{iterator.NewOrderedWeightedLines([]graph.WeightedLine{
-			multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
-			multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
-			multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
-		})},
-		want: []graph.WeightedLine{
-			multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
-			multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
-			multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
-		},
-	},
-	{
-		name: "explicit edges",
-		lines: iterator.NewOrderedWeightedLines([]graph.WeightedLine{
-			multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
-			multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
-			multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
-		}),
-		want: []graph.WeightedLine{
-			multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
-			multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
-			multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
-		},
-	},
-}
-
 type basicWeightedLines struct {
 	graph.WeightedLines
 }
 
 func TestWeightedLinesOf(t *testing.T) {
+	weightedLinesOfTests := []struct {
+		name  string
+		lines graph.WeightedLines
+		want  []graph.WeightedLine
+	}{
+		{
+			name:  "nil",
+			lines: nil,
+			want:  nil,
+		},
+		{
+			name:  "empty",
+			lines: graph.Empty,
+			want:  nil,
+		},
+		{
+			name:  "no edges",
+			lines: iterator.NewOrderedWeightedLines(nil),
+			want:  nil,
+		},
+		{
+			name: "no slice method",
+			lines: basicWeightedLines{iterator.NewOrderedWeightedLines([]graph.WeightedLine{
+				multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
+				multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
+				multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
+			})},
+			want: []graph.WeightedLine{
+				multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
+				multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
+				multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
+			},
+		},
+		{
+			name: "explicit edges",
+			lines: iterator.NewOrderedWeightedLines([]graph.WeightedLine{
+				multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
+				multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
+				multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
+			}),
+			want: []graph.WeightedLine{
+				multi.WeightedLine{F: multi.Node(-1), T: multi.Node(0), W: 1, UID: -1},
+				multi.WeightedLine{F: multi.Node(1), T: multi.Node(2), W: 2, UID: 0},
+				multi.WeightedLine{F: multi.Node(3), T: multi.Node(4), W: 3, UID: 1},
+			},
+		},
+	}
+
 	for _, test := range weightedLinesOfTests {
 		got := graph.WeightedLinesOf(test.lines)
 		if !reflect.DeepEqual(got, test.want) {

--- a/stat/samplemv/sample_test.go
+++ b/stat/samplemv/sample_test.go
@@ -163,11 +163,12 @@ func TestMetropolisHastings(t *testing.T) {
 	compareNormal(t, target, batch, nil, 5e-1, 5e-1)
 }
 
-// randomNormal constructs a random Normal distribution.
+// randomNormal constructs a random Normal distribution using the provided
+// random source.
 func randomNormal(dim int, src *rand.Rand) (*distmv.Normal, bool) {
 	data := make([]float64, dim*dim)
 	for i := range data {
-		data[i] = rand.Float64()
+		data[i] = src.Float64()
 	}
 	a := mat.NewDense(dim, dim, data)
 	var sigma mat.SymDense
@@ -180,6 +181,8 @@ func randomNormal(dim int, src *rand.Rand) (*distmv.Normal, bool) {
 }
 
 func compareNormal(t *testing.T, want *distmv.Normal, batch *mat.Dense, weights []float64, meanTol, covTol float64) {
+	t.Helper()
+
 	dim := want.Dim()
 	mu := want.Mean(nil)
 	var sigma mat.SymDense
@@ -224,10 +227,11 @@ func TestMetropolisHastingser(t *testing.T) {
 		{3, 103, 11, 51},
 		{3, 103, 51, 11},
 	} {
+		src := rand.New(rand.NewSource(1))
 		dim := test.dim
 
 		initial := make([]float64, dim)
-		target, ok := randomNormal(dim, nil)
+		target, ok := randomNormal(dim, src)
 		if !ok {
 			t.Fatal("bad test, sigma not pos def")
 		}
@@ -239,7 +243,7 @@ func TestMetropolisHastingser(t *testing.T) {
 
 		// Test the Metropolis Hastingser by generating all the samples, then generating
 		// the same samples with a burnin and rate.
-		src := rand.New(rand.NewSource(1))
+		src = rand.New(rand.NewSource(1))
 		proposal, ok := NewProposalNormal(sigmaImp, src)
 		if !ok {
 			t.Fatal("bad test, sigma not pos def")

--- a/unit/unit_test.go
+++ b/unit/unit_test.go
@@ -185,20 +185,19 @@ func TestDimensionEquality(t *testing.T) {
 	}
 }
 
-var operationTests = []struct {
-	recvOp func(Uniter) *Unit
-	param  Uniter
-	want   Uniter
-}{
-	{Dimless(1).Unit().Add, Dimless(2), Dimless(3)},
-	{Dimless(1).Unit().Mul, Dimless(2), Dimless(2)},
-	{Dimless(1).Unit().Mul, Length(2), Length(2)},
-	{Length(1).Unit().Mul, Dimless(2), Length(2)},
-	{Dimless(1).Unit().Div, Length(2), New(0.5, Dimensions{LengthDim: -1})},
-	{Length(1).Unit().Div, Dimless(2), Length(0.5)},
-}
-
 func TestOperations(t *testing.T) {
+	operationTests := []struct {
+		recvOp func(Uniter) *Unit
+		param  Uniter
+		want   Uniter
+	}{
+		{Dimless(1).Unit().Add, Dimless(2), Dimless(3)},
+		{Dimless(1).Unit().Mul, Dimless(2), Dimless(2)},
+		{Dimless(1).Unit().Mul, Length(2), Length(2)},
+		{Length(1).Unit().Mul, Dimless(2), Length(2)},
+		{Dimless(1).Unit().Div, Length(2), New(0.5, Dimensions{LengthDim: -1})},
+		{Length(1).Unit().Div, Dimless(2), Length(0.5)},
+	}
 	t.Parallel()
 	for i, test := range operationTests {
 		var got Uniter

